lib/streamlit/runtime/memory_media_file_storage.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2018-2022 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""MediaFileStorage implementation that stores files in memory."""
+
+import contextlib
+import hashlib
+import mimetypes
+import os.path
+from typing import Union, NamedTuple, Dict, Optional, List
+
+from streamlit.logger import get_logger
+from streamlit.runtime.media_file_storage import (
+    MediaFileStorage,
+    MediaFileStorageError,
+    MediaFileKind,
+)
+from streamlit.runtime.stats import CacheStatsProvider, CacheStat
+
+LOGGER = get_logger(__name__)
+
+PREFERRED_MIMETYPE_EXTENSION_MAP = {
+    "image/jpeg": ".jpeg",
+    "audio/wav": ".wav",
+}
+
+
+def _calculate_file_id(
+    data: bytes, mimetype: str, filename: Optional[str] = None
+) -> str:
+    """Hash data, mimetype, and an optional filename to generate a stable file ID.
+
+    Parameters
+    ----------
+    data
+        Content of in-memory file in bytes. Other types will throw TypeError.
+    mimetype
+        Any string. Will be converted to bytes and used to compute a hash.
+    filename
+        Any string. Will be converted to bytes and used to compute a hash.
+    """
+    filehash = hashlib.new("sha224")
+    filehash.update(data)
+    filehash.update(bytes(mimetype.encode()))
+
+    if filename is not None:
+        filehash.update(bytes(filename.encode()))
+
+    return filehash.hexdigest()
+
+
+def get_extension_for_mimetype(mimetype: str) -> str:
+    # Python mimetypes preference was changed in Python versions, so we specify
+    # a preference first and let Python's mimetypes library guess the rest.
+    # See https://bugs.python.org/issue4963
+    #
+    # Note: Removing Python 3.6 support would likely eliminate this code
+    if mimetype in PREFERRED_MIMETYPE_EXTENSION_MAP:
+        return PREFERRED_MIMETYPE_EXTENSION_MAP[mimetype]
```

Comment:
I don't remember what finally got us to get rid of unofficial Python 3.6 support, but we did get rid of it at some point relatively recently, so we may be able to remove this code now

Reply:
Yes! I'll address this post-merge.